qcengine/programs/adcc.py

```diff
@@ -0,0 +1,155 @@
+"""
+Calls adcc
+"""
+import json
+import os
+import sys
+from pathlib import Path
+import importlib
+from typing import TYPE_CHECKING, Any, Dict, Optional, Tuple
+import numpy as np
+
+from qcelemental.models import AtomicResult, AtomicResultProperties, Provenance
+from qcelemental.util import deserialize, parse_version, safe_version, which, which_import
+
+from ..exceptions import InputError, RandomError, ResourceError, UnknownError
+from .model import ProgramHarness
+
+if TYPE_CHECKING:
+    from qcelemental.models import AtomicInput
+
+    from ..config import TaskConfig
+
+
+def is_module_available(module):
+    """Check using importlib if a module is available."""
+    import importlib
+
+    try:
+        mod = importlib.import_module(module)
+    except ImportError:
+        return False
+    return True
+
+
+class AdccHarness(ProgramHarness):
+
+    _defaults = {
+        "name": "adcc",
+        "scratch": False,
+        "thread_safe": False,
+        "thread_parallel": True,
+        "node_parallel": False,
+        "managed_memory": True,
+    }
+    version_cache: Dict[str, str] = {}
+
+    class Config(ProgramHarness.Config):
+        pass
+
+    @staticmethod
+    def found(raise_error: bool = False) -> bool:
+        """Whether adcc harness is ready for operation.
+        Parameters
+        ----------
+        raise_error: bool
+            Passed on to control negative return between False and ModuleNotFoundError raised.
+        Returns
+        -------
+        bool
+            If adcc is found, returns True.
+            If raise_error is False and adcc is missing, returns False.
+            If raise_error is True and adcc is missing, the error message is raised.
+        """
+
+        adcc_found = is_module_available("adcc")
+        if adcc_found:
+            return True
+        if raise_error:
+            raise ModuleNotFoundError("adcc not found.")
+        return False
+
+    def get_version(self) -> str:
+        self.found(raise_error=True)
+        import adcc
+
+        # candidate_version = self.version_cache[which_prog]
+        candidate_version = adcc.__version__
+
+        return candidate_version
+
+    def compute(self, input_model: "AtomicInput", config: "TaskConfig") -> "AtomicResult":
+        """
+        Runs adcc
+        """
+        self.found(raise_error=True)
+        import adcc
+
+        mol = input_model.molecule
+        model = input_model.model
+        kws = input_model.keywords
+        xyz = mol.to_string(dtype="xyz+", units="Bohr")
+        xyz = "\n".join(xyz.split("\n")[2:])
+        scfres = adcc.backends.run_hf(
+            backend=None,  # auto-select available backend
+            xyz=xyz,
+            multiplicity=mol.molecular_multiplicity,
+            charge=mol.molecular_charge,
+            basis=model.basis,
+            #   conv_tol=model.scf_conv,
+            #   max_iter=,
+        )
```

Comment:
Hmm. I'm not sure about this. The way we designed it this was supposed to be "internal" and for testing only. To be fair the interface is a bit too naive from my point of view to start using it in a third-party package ... of course we can flesh it out a bit and promote it to something proper if that's needed.

Reply:
Having some kind of wrapper code would be nice I guess. Otherwise, too much of the code here would be checking for program availability etc... I think either the `run_hf` needs to be improved (run from schema) or we limit ourselves to Psi4 for now. What do you think @loriab?

Reply:
I'm hoping the external HF is a medium-term solution, to be replaced (or at least alternate) by atomicresult.wavefunction and an unknown ERI transport. `import psi4` (or `import pyscf` -- no particular reason we dn't have a wrapper for that) is fine if you don't think `adcc.run_hf` is advanced enough. it'd be good to incorporate the helper program into the `found` detection to avoid aggravation. nwchem has an example as it requires networkx.

Reply:
For now I would feel more comfortable with using psi4 or pyscf, but I agree that having a nice interface to pass such data would be nice in the long run. What we have in adcc works kind of ok, but of course it requires people to actually implement it for every backend. Would be good if the community agreed on a standard.